src/som/interpreter/objectstorage/ObjectTransitionSafepoint.java

```diff
@@ -140,12 +155,27 @@ public void writeAndGeneralizeSlot(final SObject obj, final SlotDefinition slot,
     replaceAssumptionAndWaitForSafepointEnd();
   }
 
+  public void ensureSlotAllocatedToAvoidDeadlock(final SObject obj,
+      final SlotDefinition slot) {
+    waitForSafepointStart();
+
+    // Safepoint phase, used to update the object
+    // object is required to handle updates from multiple threads correctly
+    obj.ensureSlotAllocatedToAvoidDeadlock(slot);
+
+    replaceAssumptionAndWaitForSafepointEnd();
+  }
+
   private void replaceAssumptionAndWaitForSafepointEnd() {
     // update the assumption
-    synchronized (this) {
-      // might have been replaced by another thread already
-      if (!noSafePoint.isValid()) {
-        noSafePoint = create();
+    int active = transitionsInProgress.decrementAndGet();
+
+    if (active == 0) {
+      synchronized (this) {
```

Comment:
Is it on purpose that you got two separate synchronization mechanisms here?

This might be an optimization I guess, but I may be simpler to stick to exactly one (`synchronized` only?), if possible?